### PR TITLE
tls: plumb time source to context manager

### PR DESCRIPTION
**Pull request by ramaraochavali**
*Description*: Plumbs time source to context manager as discussed in https://github.com/envoyproxy/envoy/pull/4701. This is to allow control time on `Utility::getDaysUntilExpiration` method.
*Risk Level*: Low
*Testing*: Existing automated tests
*Docs Changes*: N/A
*Release Notes*: N/A

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@lizan can you PTAL?

File: test/common/ssl/utility_test.cc [hunk 1]
```diff
@@ -58,11 +59,14 @@ TEST(UtilityTest, TestGetSerialNumber) {
 TEST(UtilityTest, TestDaysUntilExpiration) {
   bssl::UniquePtr<X509> cert = readCertFromFile(
       TestEnvironment::substitute("{{ test_rundir }}/test/common/ssl/test_data/san_dns_cert.pem"));
-  EXPECT_LE(0, Utility::getDaysUntilExpiration(cert.get()));
+  Event::SimulatedTimeSystem time_source;
+  time_source.setSystemTime(std::chrono::system_clock::from_time_t(1540080000));
```
Comment (jmarantz) [inline test/common/ssl/utility_test.cc]: comment where that constant comes from  and/or declare it as a const.

File: test/common/ssl/context_impl_test.cc [hunk 2]
```diff
@@ -198,7 +203,8 @@ TEST_F(SslContextImplTest, TestGetCertInformationWithSAN) {
   MessageUtil::loadFromYaml(TestEnvironment::substitute(yaml), tls_context);
   ClientContextConfigImpl cfg(tls_context, factory_context_);
   Runtime::MockLoader runtime;
-  ContextManagerImpl manager(runtime);
+  Event::SimulatedTimeSystem time_system;
```
Comment (jmarantz) [inline test/common/ssl/context_impl_test.cc]: suggest putting this as a test class member var.

**Comment by htuch** [comment]
@ramaraochavali can you update the description to explain that it's ultimately ` Utility::getDaysUntilExpiration` we want to control the time on? Thanks.

**Comment by ramaraochavali** [comment]
@jmarantz addressed the comments. PTAL.

File: test/common/ssl/context_impl_test.cc [hunk 3]
```diff
@@ -27,7 +27,10 @@ using testing::ReturnRef;
 namespace Envoy {
 namespace Ssl {
 
-class SslContextImplTest : public SslCertsTest {};
+class SslContextImplTest : public SslCertsTest {
+public:
```
Comment (jmarantz) [inline test/common/ssl/context_impl_test.cc]: fwiw usually 'protected:' is used here as all the test methods are effectively subclasses. Just a suggestion.

File: test/common/ssl/context_impl_test.cc [hunk 4]
```diff
@@ -246,8 +244,7 @@ TEST_F(SslContextImplTest, TestNoCert) {
   Json::ObjectSharedPtr loader = TestEnvironment::jsonLoadFromString("{}");
   ClientContextConfigImpl cfg(*loader, factory_context_);
   Runtime::MockLoader runtime;
-  Event::SimulatedTimeSystem time_system;
-  ContextManagerImpl manager(runtime, time_system);
+  ContextManagerImpl manager(runtime, time_system_);
```
Comment (jmarantz) [inline test/common/ssl/context_impl_test.cc]: I guess while in here I would also put runtime and manager in the class to factor out these common initializations but this is pre-existing, so I will leave it up to you.

Comment (ramaraochavali) [reply test/common/ssl/context_impl_test.cc]: I will do a follow-up PR with this

**Pull request merged**